lib/charms/istio_k8s/v0/istio_info.py

```diff
@@ -0,0 +1,239 @@
+"""TODO: Add a proper docstring here.
+
+* using requirer:
+    * handle info changed event  <-- suggest a charm test to confirm this?
+
+* using provider:
+    * handle send error
+    * use the is_ready()
+
+
+requires cosl
+"""
+from typing import List, Optional
+
+from cosl.interfaces.utils import DatabagModelV2, DataValidationError
+from ops import CharmBase, Object, EventBase, CharmEvents, RelationDataAccessError, EventSource
+from pydantic import BaseModel, Field
+
+# The unique Charmhub library identifier, never change it
+LIBID = "17dff53d5bf649f29614365bde32451b"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+DEFAULT_RELATION_NAME = "istio-info"
+
+
+# Interface schema
+
+class IstioInfoAppData(DatabagModelV2, BaseModel):
+    """Data model for the istio-info interface."""
+
+    # TODO: Confirm that comparison between these objects works as expected so is_ready below works.
+
+    root_namespace: str = Field(
+        description="The root namespace for the Istio installation.",
+        examples=["istio-system"],
+    )
+
+    # To make it sortable, which is useful for comparing unordered sets like how relation data is stored.
+    # TODO: Move this to DatabagModelV2?
+    def __lt__(self, other):
+        return tuple(self.model_dump().values()) < tuple(other.model_dump().values())
+
+    def __le__(self, other):
+        return tuple(self.model_dump().values()) <= tuple(other.model_dump().values())
+
+    def __gt__(self, other):
+        return tuple(self.model_dump().values()) > tuple(other.model_dump().values())
+
+    def __ge__(self, other):
+        return tuple(self.model_dump().values()) >= tuple(other.model_dump().values())
+
+
+
+# Requirer library
+
+class DataChangedEvent(EventBase):
+    """Charm Event triggered when the istio-info relation changes."""
+
+
+class IstioInfoRequirerCharmEvents(CharmEvents):
+    """Events raised by the IstioInfoRequirer class."""
+    info_changed = EventSource(DataChangedEvent)
+
+
+class IstioInfoRequirer(Object):
+    """Class for handling the requirer side of the istio-info relation."""
+
+    on = IstioInfoRequirerCharmEvents()
+
+    def __init__(
+            self,
+            charm: CharmBase,
+            relation_name: str = DEFAULT_RELATION_NAME,
+    ) -> None:
+        super().__init__(charm, relation_name)
+
+        self._charm = charm
+        self._relation_name = relation_name
+
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_changed,
+            self.on_relation_changed
+        )
+
+    def on_relation_changed(self, _: EventBase) -> None:
+        self.on.info_changed.emit()
```

Comment:
Adding a comment here because this might raise eyebrows.  

Yes, if the charm wants to know this relation data changed it could just observe the regular `relation_changed` event directly directly.  I argue we should do this pattern instead because I argue the relation_changed event is a low-level thing this lib is responsible for controlling, and then it should emit events for the charm to observe.

...but admittedly in this case it doesn't make a functional difference, just a cognitive one

Reply:
I think that is fully reasonable, and something I've been arguing for a **long** time.

Reply:
Feels in line too with any of the more complex libraries, too, like [tls certs](https://github.com/canonical/tls-certificates-interface/blob/f54b2df50663bcc81fa157b7ee5b34086a0580c7/lib/charms/tls_certificates_interface/v4/tls_certificates.py#L969) which "operates" the relation events and emits certificate_available when something the charm should care about changes

Reply:
I am generally opposed to this model. tls certs actually does a lot of logic, but, most libraries that use these custom events do absolutely nothing other than change the hook names from relation_changed to <something_else>_changed. It seems like an obfuscation for no real benefit. It can be confusing when every lib wants me to watch a differently named event even though they are all just firing on relation_changed.

Reply:
Yeah I have sympathy for @dstathis's argument.  The two things that sway me are:
* observing the lib's event guarantees order (lib handles the relation event before the charm does).  In this case it doesn't matter (since the lib does nothing), but in general it can
* always following the pattern "lib handles relation events, lib emits events for charm to handle", even when it doesn't matter right now, means if we patch this lib in a way that *does* need ordering then it "just works" in all charms.  Otherwise, it would be hard to signal that to a charm author

Reply:
see: https://github.com/canonical/parca-k8s-operator/pull/409

Reply:
This discussion has moved to [here](https://github.com/canonical/charm-relation-building-blocks/pull/1/files#r1941781590)